### PR TITLE
Permissions & Realm routes

### PR DESCRIPTION
**Pull request by bogdanromanx**

### CODE REVIEW

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala [hunk 1]
```diff
@@ -34,22 +35,27 @@ object instances extends FailFastCirceSupport {
     Encoder.encodeString.contramap(fd => s"${fd.toMillis} ms")
 
   implicit val iamErrorEncoder: Encoder[IamError] = {
-    implicit val config = rejectionConfig
+    implicit val config: Configuration = rejectionConfig
```
Comment (hygt) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: I think you can make the config implicit in the object scope, it's only use in these methods anyway.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala [hunk 2]
```diff
@@ -129,6 +136,20 @@ object instances extends FailFastCirceSupport {
       }
     }
 
+  implicit val permissionsStatusCode: RejectionStatusCode[PermissionsRejection] =
+    new RejectionStatusCode[PermissionsRejection] {
+      override def apply(rej: PermissionsRejection): StatusCode = rej match {
```
Comment (hygt) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: If all of them should yield a `BadRequest`, we can simplify.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala [hunk 3]
```diff
@@ -129,6 +136,20 @@ object instances extends FailFastCirceSupport {
...
+        case CannotSubtractEmptyCollection          => StatusCodes.BadRequest
+        case _: CannotSubtractFromMinimumCollection => StatusCodes.BadRequest
+        case CannotSubtractFromEmptyCollection      => StatusCodes.BadRequest
+        case _: CannotSubtractUndefinedPermissions  => StatusCodes.BadRequest
+        case CannotAppendEmptyCollection            => StatusCodes.BadRequest
+        case CannotReplaceWithEmptyCollection       => StatusCodes.BadRequest
+        case CannotDeleteMinimumCollection          => StatusCodes.BadRequest
+        case _: IncorrectRev                        => StatusCodes.BadRequest
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: case _: IncorrectRev                        => StatusCodes.Conflict

Why all those cases `CannotReplaceWithEmptyCollection`, `CannotAppendEmptyCollection `, `CannotSubtractFromEmptyCollection`? This is all `StatusCodes .NotFound` to me.

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                  complete {
+                    optRev match {
+                      case Some(rev) => permissions.fetchAt(rev).runToFuture
+                      case None      => permissions.fetch.runToFuture
+                    }
+                  }
+                }
+              }
+            },
+            (put & parameter("rev".as[Long])) { rev =>
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala]: There is no `createPermissions`? (where the rev query parameter is not required?)
This is not consistent with what we decided to document: https://bluebrain.github.io/nexus/docs/api/iam/iam-permissions-api.html#create-permissions

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+              entity(as[PatchPermissions]) {
+                case Replace(set) =>
```
Comment (umbreak) [inline src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala]: In [the documentation](https://bluebrain.github.io/nexus/docs/api/iam/iam-permissions-api.html#replace-permissions) neither the create nor the replace require to have `"@type": "..."`. This was defined just for PATCH operations (Append and Subtract).

If we want to change that, we have to change it in the docs + in the ACLs routes too (in order for it to be consistent)

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala (hunk 4 above)
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/routes/PermissionsRoutes.scala]: permissions already exist by default, there is no create

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala (hunk 3 above)
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: they are part of the logic, fixed the incorrect rev

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala (hunk 2 above)
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: I think it's a good idea to keep them explicit such that when we add another one the compiler will complain...

File: src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala (hunk 1 above)
Comment (bogdanromanx) [reply src/main/scala/ch/epfl/bluebrain/nexus/iam/marshallers/instances.scala]: good point, will fix

**Pull request merged**